### PR TITLE
Advanced permissions phase 1

### PR DESCRIPTION
**Pull request by mkraft**
This is the PR to merge the "Advanced Permissions Phase 1" feature branch into master.

All the individual commits to this branch have already received a full PR review, and the branch as a whole has undergone a detailed QA pass on a spinmint, so only minimal review should be required at this stage.

### CODE REVIEW

File: app/authorization.go [hunk 1]
```diff
@@ -181,17 +181,33 @@ func (a *App) HasPermissionToChannelByPost(askingUserId string, postId string, p
...
+	}
+
+	if a.HasPermissionTo(askingUserId, model.PERMISSION_EDIT_OTHER_USERS) {
+		return true
+	}
+
+	return false
+}
+
+func (a *App) RolesGrantPermission(roleNames []string, permissionId string) bool {
```
Comment (lieut-data) [inline app/authorization.go]: nit: It was clear before from the name that this was a test, but I could read this in English as an action now, with the "roles granting permission". Not a blocker.

File: store/sqlstore/upgrade.go [hunk 2]
```diff
@@ -374,6 +374,10 @@ func UpgradeDatabaseToVersion48(sqlStore SqlStore) {
 }
 
 func UpgradeDatabaseToVersion49(sqlStore SqlStore) {
+	// This version of Mattermost includes an App-Layer migration which migrates from hard-coded roles configured by
+	// a number of parameters in `config.json` to a `Roles` table in the database. The migration code can be seen
+	// in the file `app/app.go` in the function `DoAdvancedPermissionsMigration()`.
```
Comment (lieut-data) [inline store/sqlstore/upgrade.go]: What's the context for not just doing this here, in the upgrade path?

File: app/app.go [hunk 3]
```diff
@@ -495,3 +495,57 @@ func (a *App) Handle404(w http.ResponseWriter, r *http.Request) {
 
 	utils.RenderWebAppError(w, r, err, a.AsymmetricSigningKey())
 }
+
+// This function migrates the default built in roles from code/config to the database.
+func (a *App) DoAdvancedPermissionsMigration() {
+	// If the migration is already marked as completed, don't do it again.
+	if result := <-a.Srv.Store.System().GetByName(ADVANCED_PERMISSIONS_MIGRATION_KEY); result.Err == nil {
```
Comment (lieut-data) [inline app/app.go]: Any reason not to just leverage a transaction? Or is this path expose to explicitly allow someone to clear the bit and force the migration a second time?

File: store/sqlstore/upgrade.go (hunk 2 above)
Comment (grundleborg) [reply store/sqlstore/upgrade.go]: Short answer: It's necessary to do it in the app layer to avoid circular dependencies.

File: app/app.go (hunk 3 above)
Comment (lieut-data) [reply app/app.go]: (This question implicitly assumes it makes more sense to do in `upgrade.go`, but you've answered that question. I guess my only question is whether or not it still makes sense to do it as an all-or-nothing transaction, but I do note that it seems to self-recover if a role already exists.)

File: store/sqlstore/upgrade.go (hunk 2 above)
Comment (grundleborg) [reply store/sqlstore/upgrade.go]: See discussion on the original PR that landed this here for more context: https://github.com/mattermost/mattermost-server/pull/8159/files#diff-23903225b64d2e452b1d25db40371ff3R81

File: app/app.go (hunk 3 above)
Comment (grundleborg) [reply app/app.go]: Context for this one is here: https://github.com/mattermost/mattermost-server/pull/8159/files/a14b410200182a567c053278048b1745462bcebc..ed80e66b72b7fcff4cc3ae64f15d4da9ddedb034#diff-886cebc00c073731af1b3da26f11e536R468

File: app/authorization.go (hunk 1 above)
Comment (grundleborg) [reply app/authorization.go]: I actually agree with you on this one personally, but it got changed in a previous round of review :man_shrugging:

**Pull request merged**